### PR TITLE
Fix dbt deps issues (#778 #994 #895)

### LINKED ISSUE #778: Full support for dbt deps on Windows
OSError: [WinError 17] The system cannot move the file to a different disk drive: 'C:\\Users\\UserName\\AppData\\Local\\Temp\\dbt-downloads\\75cce4837e0fd73e2934f79b9a5238a9' -> 'dbt_modules\\snowplow'

This can happen on Windows if the temp dir is on a different drive than the dbt project dir

### LINKED ISSUE #895: Dbt downloads directory same across different users
## Issue

Dbt downloads temporary directory cannot be shared between users.

### Issue description
The `dbt deps` command, creates a folder `/tmp/dbt-downloads` if it needs to download anything. See [deps.py](https://github.com/fishtown-analytics/dbt/blob/2ff26aa41a324902025a980cc839dbef41df95cf/dbt/task/deps.py#L22)

### Results
If there are more than one users that may run dbt, the folder persists and is owned by whoever created it first. Later, if another user runs dbt and dbt needs to download something, it causes a permissions error and the second user cannot use dbt in that way.

### System information
The output of `dbt --version`:
```
installed version: 0.10.1
   latest version: 0.10.1
Up to date!
```

### Steps to reproduce
If you have a `packages.yml` with
```
packages:
  - git: "https://github.com/fishtown-analytics/dbt-utils"
    revision: master
```

Just run `dbt deps` as one user and then a different user.

I recommend using something along the lines of `tempfile.mkdtemp` instead, but of course there's so many ways this can be done.

### LINKED ISSUE #994: better handling of git errors in `dbt deps`
### Feature description
`dbt deps` can fail for myriad git-related reasons (git isn't installed, the repo can't be cloned, the branch doesn't exist, etc).

dbt currently shows an error like this:
```
$ dbt deps
Encountered an error:
[Errno 2] No such file or directory: '/var/folders/ql/lb0scpz112lfpcq497_mm3gc0000gn/T/dbt-downloads/18bf9a8e7f78e7bc3bd5414483d4e3f1'
```

whereas it should instead show the relevant git operation which failed.

The failure is usually visible in the logs, but I got the error above because of a bad spec in `packages.yml`, which was confusing.

### Who will this benefit?
folks who use packages

### PR DESCRIPTION
**Pull request by beckjake**
Fix some issues around dbt deps.

I think this fixes #778, fixes #994, fixes #895 - some of those are sort of judgement calls

### Behavior after this PR

 * When `git` is not available, dbt is more helpful now
```
> dbt deps
Encountered an error:
Could not find command, ensure it is in the user's PATH: "git"
```

 * When a repository is incorrect, shows the command that failed (it'd be nice to remove the username/password prompt!):
```
> dbt deps
Username for 'https://github.com': beckjake
Password for 'https://beckjake@github.com':
Encountered an error:
Got a non-zero returncode running: ['git', 'clone', '--depth', '1', 'https://github.com/fishtown-analytics/dbt-utilz.git', '23da7ac89d7d5c4086737ad8dc02feff']
```

 * When a revision is invalid, shows the ref that could not be found and the package URL:
```
> dbt deps
Encountered an error:
Error checking out spec='0.9.17' for repo https://github.com/fishtown-analytics/dbt-utils.git
fatal: Couldn't find remote ref 0.9.17
```

 * Uses `shutil.move` instead of `os.rename` to move things out of the download directory, to handle the cross-drive issue

 * By default, downloads go to a temp directory `dbt-downloads-xxxxx` where `xxxxx` is some random stuff, created by `tempfile.mkdtemp` that is removed via `shutil.rmtree` after a successful run. This should resolve the concurrent users issue. Users can override this behavior to download to a fixed directory using `DBT_DOWNLOADS_DIR`, in which case dbt will not `shutil.rmtree` the directory.

* dbt no longer ignores return codes and instead raises exceptions that have stderr/stdout

* More explicitly log what failed and where on permissions-related execution failures, this should be especially more helpful on windows

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,111 +1,82 @@
 version: 2
 jobs:
   unit:
-    docker: &py36
-      - image: python:3.6
+    docker: &containers
+      - image: fishtownjacob/test-container
```
Comment (cmcarthur) [inline .circleci/config.yml]: is this built by `Dockerfile` in this repo?

Comment (beckjake) [reply .circleci/config.yml]: Yeah. I use this gross and lame shell script to do it:
```bash
#!/bin/bash
set -xeuo pipefail
commit_message="$1"
shift
if [[ -z ${commit_message} ]]; then
    echo 'no commit message'
    exit 1
fi
docker build --tag 'test-container:latest' .
container_id="$(docker run -it test-container:latest hostname | tr -d '[[:space:]]')"
docker commit -a 'Jacob Beck <jake@fishtownanalytics.com>' -m "${commit_message}" ${container_id} test-container
docker tag test-container:latest fishtownjacob/test-container:latest
docker push fishtownjacob/test-container:latest
```

**Comment by drewbanin** [comment]
This looks really great @beckjake!

Can you think of a good way to show a more obvious error message if `git` is not in the user's path? My guess is that if someone is just getting started with dbt and they don't have git installed on windows, this message might be a little confusing:
> Could not find command, ensure it is in the user's PATH: "git"

Ideally, I think we'd show:
> Could not find command, ensure it is in the user's PATH: "git"
> Make sure `git` is installed on your machine. More information: https://docs.getdbt.com/docs/package-management

^ We can make a section about the requirements for `dbt deps` (including more info about how to install git) and link to that specific section.



I'm a little confused that you're seeing the Username/Password prompt here:
```
> dbt deps
Username for 'https://github.com': beckjake
Password for 'https://beckjake@github.com':
...
```

 I couldn't reproduce that locally -- do you think it's a git version thing? Or something else?

**Comment by beckjake** [comment]
for the missing git error, I can easily change it to this:
```
Make sure git is installed on your machine. More information: https://docs.getdbt.com/docs/package-management
Encountered an error:
Could not find command, ensure it is in the user's PATH: "git"
```

Which is what you suggested but with order reversed. Is that ok?


I think the password prompt is a global git settings thing, or maybe version-dependent. it happens to me a lot when cloning a private repository from github via `https` urls. I've always taken it as a sign that my ssh key is somehow wrong.

**Comment by atharvai** [comment]
Drew asked me to test this out on Windows, here are the results: https://gist.github.com/atharvai/a302f1777a1f5b0202dffca937cb8de3

Summary: If the `DBT_DOWNLOADS_DIR` or default temp dir are on same drive as dbt project, then `dbt deps` is successful. Otherwise it fails on `Access Denied` on `os.remove`

**Comment by drewbanin** [comment]
Thanks @atharvai! This is super helpful :)

**Comment by beckjake** [comment]
With that most recent push I believe it is now fixed. I tested on Windows with an old drive I found, and cross drive moves from both temp directories and explicit DBT_DOWNLOADS_DIR settings are working as they should.

**Comment by atharvai** [comment]
I tested the latest commit and can confirm all is good! 
One suggestion is to log if `DBT_DOWNLOADS_DIR` is set or not and value if set. It would help debugging.

Alternatively, set this log line to print full path `2018-11-14 18:06:47,742 (MainThread): STDERR: "Cloning into '0a6a5f508a311be0d4ba05de66b74124'...`

**Comment by drewbanin** [comment]
@beckjake @atharvai awesome!! Really glad we got this working. Thanks for your insight @atharvai :)

This LGTM when the tests all pass!

**Pull request merged**